airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -89,41 +102,89 @@ public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws
+          try {
+            while (!cancelled.get() && !destination.isFinished()) {
+              final Optional<AirbyteMessage> messageOptional = destination.attemptRead();
+              if (messageOptional.isPresent()) {
+                LOGGER.info("state in DefaultReplicationWorker from Destination: {}", messageOptional.get());
+                destinationMessageTracker.accept(messageOptional.get());
+              }
+            }
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+
+        final Future<?> destinationThreadExecution = executorService.submit(destinationOutputThread);
+        final Future<?> sourceThreadExecution = executorService.submit(replicationThread);
+
+        LOGGER.info("Waiting for source thread to join.");
+        sourceThreadExecution.get();
+        LOGGER.info("Source thread complete.");
+        LOGGER.info("Waiting for destination thread to join.");
+        destinationThreadExecution.get();
+        LOGGER.info("Destination thread complete.");
+
+      } catch (Exception e) {
+        hasFailed.set(true);
+        LOGGER.error("Sync worker failed.", e);
+      } finally {
+        executorService.shutdownNow();
       }
 
-      final StandardSyncSummary summary = new StandardSyncSummary()
-          .withStatus(cancelled.get() ? Status.FAILED : Status.COMPLETED)
-          .withRecordsSynced(messageTracker.getRecordCount())
-          .withBytesSynced(messageTracker.getBytesCount())
+      final ReplicationStatus outputStatus =
+          cancelled.get() ? ReplicationStatus.CANCELLED : hasFailed.get() ? ReplicationStatus.FAILED : ReplicationStatus.COMPLETED;
+      final ReplicationAttemptSummary summary = new ReplicationAttemptSummary()
+          .withStatus(outputStatus)
+          .withRecordsSynced(sourceMessageTracker.getRecordCount())
+          .withBytesSynced(sourceMessageTracker.getBytesCount())
           .withStartTime(startTime)
           .withEndTime(System.currentTimeMillis());
 
       LOGGER.info("sync summary: {}", summary);
 
-      final StandardSyncOutput output = new StandardSyncOutput()
-          .withStandardSyncSummary(summary)
+      final ReplicationOutput output = new ReplicationOutput()
+          .withReplicationAttemptSummary(summary)
           .withOutputCatalog(destinationConfig.getCatalog());
 
-      messageTracker.getOutputState().ifPresent(capturedState -> {
-        final State state = new State()
-            .withState(capturedState);
+      if (destinationMessageTracker.getOutputState().isPresent()) {
+        LOGGER.info("State capture: Updated state to: {}", destinationMessageTracker.getOutputState());
+        final State state = new State().withState(destinationMessageTracker.getOutputState().get());
         output.withState(state);
-      });
+      } else if (syncInput.getState() != null) {
+        LOGGER.warn("State capture: No new state, falling back on input state: {}", syncInput.getState());
```

Comment:
is this branch basically for backwards compatibility? If so, should we also check if the sync is successful?

Reply:
Not really. Basically we don't want to "forget" state, so if we fail to checkpoint at all we just return the initial state because we know that's true.

I need to handle the backwards compatibility case separately. I think that's going to be in here with the logic you described. If we are successful, got no output from the destination but do have output from the source, we will fall back on that.

Reply:
I see Sherif's point - is the destination emitting state an invariant? Should we fail loudly if this isn't the case?

Reply:
NVM, I just read the `PartialSuccessTemporalWorkflow` class - I guess this is something we are expecting.

To make sure I understand - when will this happen? If a destination isn't migrated over or if a destination fails before it makes it to a checkpoint?

Reply:
I actually think i want to skip backwards compatibility and just upgrade all destinations. Mainly I don't think there's a safe way to get out of the backwards compatibility state since we don't have connector - core compatibility safeties. so might as well just make a clean cut and tell people to move to the bleeding edge. #alpha

Reply:
thought more about this - emitting state should not be an invariant. full refresh syncs don't need to output state.

Reply:
> This being said we aren't catching the case where the destination is e.g: buggy and not outputting state. I don't think we want to fail the sync in that case but it is a bug that could go unnoticed

@sherifnada could you say a bit more here? It seems to me that if a destination is failing to output state when it is expected to that we _should_ fail the sync. If the destination doesn't tell us what records then we can never move state forward. Destination has to be source of truth here. So I think the new contract is that (at least on incremental--although you could make a coherent argument for FR too) that if the destination doesn't output state at least once something has gone wrong.

Typing this out it makes me realize that it would actually probably be helpful for us to track whether the source ever emitted state, so in these failure cases we at least have logs that give us a clue if the destination failed to emit state because of something with it (or the worker) or because the source never emitted any state.

Reply:
Agreed with the bit about tracking source emitting state. The invariants are: 
1. The destination should output a non-zero number of state messages if the source output any state messages
2. The destination should output state messages in monotonically increasing order (depending on order received from the source)
3. The last state message output by the destination in a successful sync should == the last state message output by the source